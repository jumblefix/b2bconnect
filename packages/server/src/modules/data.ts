export const bookData = {
  title: 'My Book',
  coverImage: 'my-book.jpg',
  isbn: '1233434234',
  description:
    'Right now when I run npm test the output is just: ' +
    'and the number of passed and failed tests (when the tests are successful). ' +
    'I would like the output to include  and any other test descriptions' +
    ' (like how the output looks when an test is run).',
  rating: 0,
  price: 66.0,
  offerPrice: 56.0,
  categoryId: 'invalid',
  publishedYear: 1999,
};

export const user = {
  name: 'example',
  email: 'example@email.com',
  password: '123456',
};
