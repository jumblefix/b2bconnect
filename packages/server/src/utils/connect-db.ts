// tslint:disable-next-line:no-var-requires
const iconv = require('iconv-lite');
// tslint:disable-next-line:no-var-requires
const encodings = require('iconv-lite/encodings');
iconv.encodings = encodings;

import { createConnection, getConnectionOptions } from 'typeorm';

export const connectDb = async () => {
  const connectionOptions = await getConnectionOptions(process.env.NODE_ENV);

  return createConnection({
    ...connectionOptions,
    name: 'default',
  });
};

export const connectDbTest = async (resetDB: boolean = false) => {
  const connectionOptions = await getConnectionOptions(process.env.NODE_ENV);

  return createConnection({
    ...connectionOptions,
    name: 'default',
    synchronize: resetDB,
    dropSchema: resetDB,
  });
};
