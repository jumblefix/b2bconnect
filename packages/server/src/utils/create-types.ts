import { generateNamespace } from '@gql2ts/from-schema';
import * as fs from 'fs';
import * as path from 'path';
import { genSchema } from './schema-utils';

const typescriptTypes = generateNamespace('GQL', genSchema());

fs.writeFile(
  path.join(__dirname, '../typings/GQL/index.d.ts'),
  typescriptTypes,
  err => {
    console.error(err);
  },
);
