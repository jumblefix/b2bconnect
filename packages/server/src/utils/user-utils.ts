import { Request, Response } from 'express';
import * as IORedis from 'ioredis';
import { v4 } from 'uuid';
import { ValidationError } from 'yup';
import {
  FORGOT_PASSWORD_PREFIX,
  REDIS_SESSION_PREFIX,
  TokenTypes,
  USER_SESSION_PREFIX,
} from '../constants';
import { User } from '../entity/User';
import { GraphQLMiddlewareFunc, Resolver } from '../typings/app-utility-types';

export const createForgotPasswordLink = async (
  url: string,
  userId: string,
  redis: IORedis.Redis,
) => {
  const id = v4();
  await redis.set(`${FORGOT_PASSWORD_PREFIX}${id}`, userId, 'ex', 60 * 60);
  return `${url}/change-password/${id}`;
};

export const removeAllUsersSessions = async (
  userId: string,
  redis: IORedis.Redis,
) => {
  const sessionIds = await redis.lrange(
    `${USER_SESSION_PREFIX}${userId}`,
    0,
    -1,
  );

  const promises = [];
  // tslint:disable-next-line:prefer-for-of
  for (let i = 0; i < sessionIds.length; i += 1) {
    promises.push(redis.del(`${REDIS_SESSION_PREFIX}${sessionIds[i]}`));
  }
  await Promise.all(promises);
};

export const formatYupError = (err: ValidationError) => {
  const errors: [{path: string; message: string}] = [] as any;
  err.inner.forEach(e => {
    errors.push({
      path: e.path,
      message: e.message,
    });
  });

  return errors;
};

export const createMiddleware = (
  middlewareFunc: GraphQLMiddlewareFunc,
  resolverFunc: Resolver,
) => (parent: any, args: any, context: any, info: any) =>
  middlewareFunc(resolverFunc, parent, args, context, info);

export const middleware = async (
  resolver: Resolver,
  parent: any,
  args: any,
  context: any,
  info: any,
) => {
  return resolver(parent, args, context, info);
};

export const createTokenLink = async (
  url: string,
  userId: string,
  redis: IORedis.Redis,
  type: TokenTypes,
) => {
  const id = v4();
  await redis.set(id, userId, 'ex', 60 * 60 * 24);
  return `${url}/${type}/${id}`;
};

export const confirmEmail = async (req: Request, res: Response) => {
  const { id } = req.params;
  const RedisInstance = new IORedis();
  const userId = await RedisInstance.get(id);
  if (userId) {
    await User.update({ id: userId }, { confirmed: true });
    await RedisInstance.del(id);
    res.redirect(`${process.env.FRONTEND_HOST}/login`);
  } else {
    res.json({ message: 'invalid or expired register token' });
  }
};

export const resetPassword = async (req: Request, res: Response) => {
  const { id } = req.params;
  const RedisInstance = new IORedis();
  const userId = await RedisInstance.get(id);
  if (userId) {
    res.redirect(`${process.env.FRONTEND_HOST}/reset/${id}`);
  } else {
    res.json({ message: 'invalid or expired reset token' });
  }
};
