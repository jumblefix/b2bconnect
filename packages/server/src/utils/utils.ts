import {
  formatError as formatApolloError,
  isInstance as isApolloErrorInstance,
} from 'apollo-errors';
import * as nodemailer from 'nodemailer';
import slugify from 'slugify';
import { ValidationError } from 'yup';
import { Env } from '../constants';
import {
  AuthenticationError,
  AuthorizationError,
  InputValidationError,
} from './errors';

export const add = (...args: number[]) => args.reduce((c, p) => c + p);

export const formatYupError = (err: ValidationError) => {
  const errors: [{ path: string; message: string }] = [] as any;
  err.inner.forEach(e => {
    errors.push({
      path: e.path,
      message: e.message,
    });
  });
  return errors;
};

export function formatError(error: any) {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    // log internalData to stdout but not include it in the formattedError
    console.error(
      JSON.stringify({
        type: `error`,
        data: originalError.data,
        internalData: originalError.internalData,
      }),
    );
  }
  return formatApolloError(error);
}

export const makeSlug = (str: string) => slugify(str, { lower: true });

export const sendEmail = async ({
  to,
  subject,
  text,
  html,
}: SendEmailParams) => {
  let mailConfig: any;
  if (process.env.NODE_ENV === Env.production) {
    mailConfig = {
      host: 'smtp.sendgrid.net',
      port: 587,
      auth: {
        user: 'real.user',
        pass: 'verysecret',
      },
    };
  } else {
    mailConfig = {
      host: process.env.MAIL_HOST,
      port: process.env.MAIL_PORT,
      auth: {
        user: process.env.MAIL_USER,
        pass: process.env.MAIL_PASS,
      },
    };
  }

  const transport = nodemailer.createTransport(mailConfig);

  try {
    const info = await transport.sendMail({
      to,
      subject,
      text,
      html,
    });

    console.log(nodemailer.getTestMessageUrl(info));
  } catch (error) {
    console.error(error);
    throw new Error(error);
  }
  return true;
};

export const checkAdminRights = (session: any) => {
  const { userId, isAdmin } = session;
  if (!userId) {
    throw new AuthenticationError();
  }
  if (!isAdmin) {
    throw new AuthorizationError();
  }
};

export const validateInputs = async (schema: any, inputs: any) => {
  try {
    await schema.validate(inputs, { abortEarly: false });
  } catch (err) {
    const errors = formatYupError(err);
    throw new InputValidationError({
      data: errors,
    });
  }
};
