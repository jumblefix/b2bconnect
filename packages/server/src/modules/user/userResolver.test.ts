import { graphql } from 'graphql';
import * as ioredis from 'ioredis';
import { Connection } from 'typeorm';
import { connectDbTest } from '../../utils/connect-db';
import { createDb } from '../../utils/create-db';
import { genSchema } from '../../utils/schema-utils';
import { user } from '../data';

let connection: Connection;
let registerId: any = '';
let redis: ioredis.Redis;

beforeAll(async () => {
  await createDb();
  redis = new ioredis();
  connection = await connectDbTest(true);
});

afterAll(async () => {
  connection.close();
});

const registerInvalidData = {
  caseId: 'register with invalid data',
  query: `mutation {
             register(name:"", email: "dummy.com", password: "12") {
               email
               name
               id
} }`,
  session: { userId: '' },
};

const registerAdminValidData = {
  caseId: 'register admin with valid data w/o login',
  query: `mutation {
             register(name:"admin user", email: "dummy@dummy.com", password: "123456", admin:true) {
               email
               name
               id }}`,
  session: { userId: '' },
};

const registerUserValidData = {
  caseId: 'registerUserValidData',
  query: `mutation {
           register(name:"${user.name}", email: "${user.email}", password: "${
    user.password
  }") { email name id}}`,
  session: { userId: '' },
};

const reRegister = {
  caseId: 'reRegister',
  query: `mutation {
             register(name:"${user.name}", email: "${user.email}", password: "${
    user.password
  }") {email name }}`,
  session: { userId: '' },
};

const login = {
  caseId: 'login',
  query: `mutation {
         login( email: "${user.email}", password: "${
    user.password
  }") { email name } }`,
  session: { userId: '' },
};

const invalidLogin = {
  caseId: 'invalidLogin',
  query: `mutation {
         login( email: "${user.email}", password: "invalid") { email name } }`,
  session: { userId: '' },
};

const invalidUser = {
  caseId: 'invalidUser',
  query: `mutation { login(email: "nonexisting@email.com", password: "123456") { name } }`,
  session: { userId: '' },
};

const getUser = {
  caseId: 'getUser',
  query: `{ getUser(id:"123"){ id }}`,
  session: { userId: '' },
};

const me = (id: string): TestCase => {
  return {
    caseId: 'me',
    query: `query{ me{ name } }`,
    session: { userId: id },
  };
};

const logout = (id: string): TestCase => {
  return {
    caseId: 'logout',
    query: `mutation{ logout }`,
    session: { userId: id },
  };
};

const meLoggedOut = {
  caseId: 'meLoggedOut',
  query: `mutation{ logout }`,
  session: { userId: '' },
};

describe('user resolver', () => {
  const cases = [
    registerInvalidData,
    registerAdminValidData,
    registerUserValidData,
    reRegister,
    login,
    invalidLogin,
    invalidUser,
    getUser,
    me,
    logout,
    meLoggedOut,
  ];

  cases.forEach(c => {
    it(`case:`, async () => {
      const { query, session, caseId } =
        typeof c === 'function' ? c(registerId) : c;

      const context = {
        redis,
        session,
        db: connection,
        url: 'http://localhost:4000',
        req: {},
      };

      const loggedInContext = (userId: any) => {
        return {
          redis,
          db: connection,
          session: {
            userId,
            destroy: () => null,
          },
          url: 'http://localhost:4000',
          req: {
            sessionID: '123',
          },
        };
      };

      const ctx = session.userId ? loggedInContext(session.userId) : context;
      const result: any = await graphql(genSchema(), query, null, ctx, {});

      if (caseId === 'registerUserValidData') {
        const { data } = result;
        registerId = data.register!.id;
        expect(data.register!.name).toEqual(user.name);
      } else {
        expect(result).toMatchSnapshot();
      }
    });
  });
});
