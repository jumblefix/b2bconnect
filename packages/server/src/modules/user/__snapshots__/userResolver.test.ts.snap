// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user resolver case: 1`] = `
Object {
  "data": Object {
    "register": null,
  },
  "errors": Array [
    [GraphQLError: Validation failed],
  ],
}
`;

exports[`user resolver case: 2`] = `
Object {
  "data": Object {
    "register": null,
  },
  "errors": Array [
    [GraphQLError: Please login to continue],
  ],
}
`;

exports[`user resolver case: 3`] = `
Object {
  "data": Object {
    "register": null,
  },
  "errors": Array [
    [GraphQLError: example@email.com is already registered with us],
  ],
}
`;

exports[`user resolver case: 4`] = `
Object {
  "data": Object {
    "login": Object {
      "email": "example@email.com",
      "name": "example",
    },
  },
}
`;

exports[`user resolver case: 5`] = `
Object {
  "data": Object {
    "login": null,
  },
  "errors": Array [
    [GraphQLError: Invalid email or password],
  ],
}
`;

exports[`user resolver case: 6`] = `
Object {
  "data": Object {
    "login": null,
  },
  "errors": Array [
    [GraphQLError: User not found],
  ],
}
`;

exports[`user resolver case: 7`] = `
Object {
  "data": Object {
    "getUser": null,
  },
}
`;

exports[`user resolver case: 8`] = `
Object {
  "data": Object {
    "me": Object {
      "name": "example",
    },
  },
}
`;

exports[`user resolver case: 9`] = `
Object {
  "data": Object {
    "logout": true,
  },
}
`;

exports[`user resolver case: 10`] = `
Object {
  "data": Object {
    "logout": false,
  },
}
`;
